Support net_mgmt events from userspace

jukkar opened the pull request:
Instead of modifiying net_mgmt API directly for userspace support, create a AF_NET_MGMT socket address family and use the socket API to receive network management events in userspace application. This version only implements the event listening support, but in the future we can add a way to set values in the system via network management socket implementation.

jukkar committed:
net: mgmt: Add info length to event wait API

The info parameter is difficult to use if the caller does not
get information how long the info struct is. So add info_length
parameter to net_mgmt_event_wait_on_iface() and
net_mgmt_event_wait() APIs.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: sockets: mgmt: Add AF_NET_MGMT address family support

Allow application to listen network management events using
BSD socket API. Application needs to create the socket using
AF_NET_MGMT address family. At this point we only support
receiving network management events that the network subsystem
is sending.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: sock: mgmt: Add network management socket app

A simple application that listens network management events
using BSD socket API.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

ioannisg commented:
At least a commit indicates that this is a bug-fix, @jukkar , so I wonder if this needs to be backported to 1.14.

dbkinder reviewed (approved):
doc changes LGTM, thanks.

andrewboie commented on samples/net/sockets/net_mgmt/sample.yaml:
```diff
@@ -0,0 +1,15 @@
+tests:
+  sample.net.sockets.net_mgmt.kernelmode:
+    extra_configs:
+      - CONFIG_USERSPACE=n
```
userspace isn't enabled by default

jukkar replied:
Sure. I placed the config there as one cannot leave the test "section" empty, it needs to have some setting in it.

jukkar commented:
> At least a commit indicates that this is a bug-fix, @jukkar , so I wonder if this needs to be backported to 1.14.

No, this is not a bug fix but a new feature and for 2.0 only.

jukkar commented on include/net/socket_net_mgmt.h:
```diff
@@ -0,0 +1,108 @@
+ */
+struct net_mgmt_msghdr
+{
+	/** Network manamgement version */
```
Note to self, fix typo

jukkar commented on include/net/socket_net_mgmt.h:
```diff
@@ -0,0 +1,108 @@
+ *  0x001 : The net_mgmt event info message follows directly
+ *          after the header.
+ */
+#define NET_MGMT_VERSION_1 0x0001
```
Only one version define needed, I will remove the one of them.

jukkar commented on subsys/net/lib/sockets/sockets_net_mgmt.c:
```diff
@@ -0,0 +1,339 @@
+	}
+
+	if (nm_addr) {
+		nm_addr->nm_ifindex = net_if_get_by_iface(iface);
```
Check if the iface was set and is not null

jukkar reviewed (commented):
Some notes to self to fix.

jukkar committed:
net: mgmt: Send event with info when applicaple

If the net_mgmt event has some info, like IP address, that
could be sent, then send it the same time. This is very useful
for the receiver of the event in order to know that is happening
in the system.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: utils: Add userspace support to net_addr_ntop/pton()

Allow userspace application call net_addr_ntop() and
net_addr_pton() functions.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
tests: net: utils: Add userspace tests for net_addr_ntop/pton()

Add unit tests to net_addr_ntop() and net_addr_pton() in order to
verify that the functions can be called from userspace.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: sock: mgmt: Print IP addresses for each event

Print IP address for added or removed IP address. This helps
to understand what IP address is being removed or added.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
tests: net: mgmt: Verify that info_length is correct

If info is sent by net_mgmt, make sure it contains correct data.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

pfalcon reviewed (dismissed):
>> At least a commit indicates that this is a bug-fix, @jukkar , so I wonder if this needs to be backported to 1.14.

> No, this is not a bug fix but a new feature and for 2.0 only.

I guess @ioannisg meant commit "net: mgmt: Fix the layer and layer code matching", which **is** bugfix, and has nothing to do with these pseudo-netlink sockets. So, let me do the right thing and -1 this, requesting splitting this to a few (2 for now) easily reviewable PRs, some of which actually require backporting.

jukkar commented:
I will reorganize the commits and split the PR after @tbursztyka has reviewed the code. Some of the fixes we might be able to backport to 1.14.

jukkar committed:
tests: net: socket: mgmt: Add unit tests for net_mgmt sockets

Almost the same as the net_mgmt socket sample but can be run by
sanitychecker.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: socket: Add userspace support to setsockopt()

Allow userspace application to call setsockopt() without crashing.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: socket: Add userspace support to getsockopt()

Allow userspace application to call getsockopt() without crashing.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: socket: mgmt: Add setsockopt() and getsockopt() support

Currently only setting and getting of Ethernet Qav options are
supported via this interface.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
tests: net: socket: mgmt: Add tests for get|setsockopt()

Test AF_NET_MGMT API getsockopt() and setsockopt() functions that
they work properly when called from kernel or from userspace.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
* resolved merge conflict
* added userspace support to generic getsockopt() and setsockopt() calls
* added support for getsockopt() and setsockopt() calls in AF_NET_MGMT
* implemented setting and getting of Ethernet Qav parameters
* added more unit tests

jukkar commented:
Rebased against latest master.

tbursztyka commented on include/net/net_mgmt.h:
```diff
@@ -140,6 +140,7 @@ struct net_mgmt_event_callback {
 
 #ifdef CONFIG_NET_MGMT_EVENT_INFO
 	const void *info;
+	size_t info_length;
```
It was by design: length does not need to be passed since for the event it must be a certain struct for the info. So length is implied by the event code.

jukkar replied:
The design looks a bit weird then. We need to have the length here, how the recipient would know the size otherwise. It is probably ok for IP address but now so much for something more complex.

tbursztyka replied:
Design was totally fine within net_mgmt api. Since for such event with an info, it is guaranteed to have that structure S etc... (Note as well this was done like this when tiny targets were a priority).

Now, in the current case of having a socket layer on top, that makes sense to have a length since that part is not going to interpret each and every commands to set the length in its own header. That would be indeed tedious. Can you change your commit message to reflect that?

pfalcon replied:
> Now, in the current case of having a socket layer on top

Don't blame sockets, blame userspace, that's what not "compatible" with fine optimizations like that. A move to introduce poorman's netlink layer wasn't exactly expected right away, but quite natural reaction to userspace hurdles...

tbursztyka replied:
Well, in fact, we could have the same thing in AF_NET_MGMT: user buffer length would need to be sized to the biggest possible info size, and up to the user - depending on event - to cast it to right struct. If buffer is too small, error would be returned. But where it is fine in the kernel to have such optimizations, I feel a bit uncomfortable to get the same in a socket api.

tbursztyka commented on subsys/net/ip/net_mgmt.c:
```diff
@@ -167,9 +167,9 @@ static inline void mgmt_run_callbacks(struct mgmt_event_entry *mgmt_event)
 
 	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&event_callbacks, cb, tmp, node) {
-		if (!(NET_MGMT_GET_LAYER(mgmt_event->event) ==
+		if (!(NET_MGMT_GET_LAYER(mgmt_event->event) &
```
Nack to that. The mask is only for command. Layer and layer code have to be an exact match.

pfalcon replied:
> The mask is only for command. Layer and layer code have to be an exact match.

Is documented properly somewhere? On my memory, this causes a stream of confusion and bugs in the code. Perhaps, besides documentation, we also should provide helper macros to do the matching, and require their usage.

tbursztyka replied:
#16133

jukkar replied:
I think we need to change this, how this works now is just counter intuitive. For example I was trying to catch these events `NET_EVENT_IPV6_ADDR_DEL |  NET_EVENT_IPV4_ADDR_ADD | NET_EVENT_IPV4_ADDR_DEL` and it was not possible to catch that. It makes no sense to have multiple callbacks just to catch that.

tbursztyka replied:
no, this won't work. It actually worked like that before and triggered bugs, too many false positive cb calls etc... And in the end, in the code side: it's ugly to mix layers and all. Let's keep it as it is.

jukkar replied:
We need to add more unit tests so the things will work as expected. This is constant battle to keep things working if we change things and do not have proper tests in place.

As I said earlier, as a net-mgmt library user I am expecting that I am able to use the sync version of the net_mgmt API to catch things properly. Currently this is not possible if I need to have multiple masks defined to get the desired events like I mentioned above.

tbursztyka replied:
sync version? you mean socket version? If the API is a 1:1 on core net_mgmt I don't see any problem.

Now if you want to modify the API that's another story. And that patch is not going to work properly for what you are trying to do.

jukkar replied:
> sync version? you mean socket version? If the API is a 1:1 on core net_mgmt I don't see any problem.

By sync version I am meaning `net_mgmt_event_wait_on_iface()` and `net_mgmt_event_wait_on_iface()`

> Now if you want to modify the API that's another story. And that patch is not going to work properly for what you are trying to do.

No, I do not want to change the API unless we have to. I am expecting that I can call either of the above functions with this mask `NET_EVENT_IPV6_ADDR_DEL | NET_EVENT_IPV4_ADDR_ADD | NET_EVENT_IPV4_ADDR_DEL` as currently it does not work.

tbursztyka replied:
It is an API change, since layer and layer code would be then interpreted as a mask. And your patch is not going to work: layer and layer code have overlapping bits. (false positives, etc...)
All of this has been discussed long time ago, and a fix would require to rework the bits fields, there was even discussion back then to extend the whole to 64 bits etc...

Not that it is not possible, but it's a different thing than just exposing net_mgmt through socket.

tbursztyka commented on subsys/net/lib/sockets/sockets_net_mgmt.c:
```diff
@@ -0,0 +1,346 @@
+	s32_t alloc_timeout;
+
+	/* net_mgmt event timeout */
+	s32_t wait_timeout;
```
from lines 79 and 80, these 2 attributes above don't seem necessary. At least not yet, they could be added if needed in the future.

jukkar replied:
I was planning to add a way to have a proper timeout when waiting events, thus the timeout value here.

jukkar commented:
I removed the controversial code from net_mgmt in order to proceed. We can test things without those but I think in the future we need to figure out some changes in order to use the net-mgmt sync interface efficiently.

jukkar commented:
@pfalcon I missed the reason why you gave -1 to this one, can you re-check?

pfalcon commented:
My comment was at https://github.com/zephyrproject-rtos/zephyr/pull/16743#pullrequestreview-248953430, hopefully it's fixed. Removing my -1, as I'm going on vacation.
